src/core/experimental/memory/memcpy.d

```diff
@@ -0,0 +1,401 @@
+            ++d;
+            ++s;
+        }
+        if (n & 0x02)
+        {
+            *(cast(ushort*) d) = *(cast(const ushort*) s);
+            d += 2;
+            s += 2;
+        }
+        if (n & 0x04)
+        {
+            *(cast(uint*) d) = *(cast(const uint*) s);
+            d += 4;
+            s += 4;
+        }
+        if (n & 0x08)
+        {
+            *(cast(ulong*) d) = *(cast(const ulong*) s);
+        }
+        return;
+    }
+    if (n <= 32)
+    {
+        float4 xmm0 = load16fSSE(s);
+        float4 xmm1 = load16fSSE(s-16+n);
+        store16fSSE(d, xmm0);
+        store16fSSE(d-16+n, xmm1);
+        return;
+    }
+    // NOTE(stefanos): I'm writing using load/storeUnaligned() but you possibly can
+    // achieve greater performance using naked ASM. Be careful that you should either use
+    // only D or only naked ASM.
+    if (n <= 64)
+    {
+        float4 xmm0 = load16fSSE(s);
+        float4 xmm1 = load16fSSE(s+16);
+        float4 xmm2 = load16fSSE(s-32+n);
+        float4 xmm3 = load16fSSE(s-32+n+16);
+        store16fSSE(d, xmm0);
+        store16fSSE(d+16, xmm1);
+        store16fSSE(d-32+n, xmm2);
+        store16fSSE(d-32+n+16, xmm3);
+        return;
+    }
+    import core.simd : void16;
+    lstore64fSSE(d, s);
+    // Requires _no_ overlap.
+    n -= 64;
+    s = s + n;
+    d = d + n;
+    lstore64fSSE(d, s);
+}
+
+/* Handle dynamic sizes > 128. `d` and `s` must not overlap.
+ */
+private void Dmemcpy_large(void* d, const(void)* s, size_t n)
+{
+    // NOTE(stefanos): Alternative - Reach 64-byte
+    // (cache-line) alignment and use rep movsb
+    // Good for bigger sizes and only for Intel.
+
+    // Align destination (write) to 32-byte boundary
+    // NOTE(stefanos): We're using SSE, which needs 16-byte alignment.
+    // But actually, 32-byte alignment was quite faster (probably because
+    // the loads / stores are faster and there's the bottleneck).
+    uint rem = cast(ulong) d & 15;
+    if (rem)
+    {
+        store16fSSE(d, load16fSSE(s));
+        s += 16 - rem;
+        d += 16 - rem;
+        n -= 16 - rem;
+    }
+
+    static string loop(string prefetchChoice)()
+    {
+        return
+        "
+        while (n >= 128)
+        {
+            // Aligned stores / writes
+            " ~ prefetchChoice ~ "(d, s);
+            d += 128;
+            s += 128;
+            n -= 128;
+        }
+        ";
+    }
+
+    if (n >= 20000)
+    {
+        mixin(loop!("lstore128fpSSE")());
+    }
+    else
+    {
+        mixin(loop!("lstore128fSSE")());
+    }
+
+    // We already have checked that the initial size is >= 128
+    // to be here. So, we won't overwrite previous data.
+    if (n != 0)
+    {
+        lstore128fSSE(d - 128 + n, s - 128 + n);
+    }
+}
+
+}
+else
+{
+
+/* Non-SIMD version
+ */
+void Dmemcpy(void* d, const(void)* s, size_t n)
+{
+    ubyte* dst = cast(ubyte*) d;
+    const(ubyte)* src = cast(const(ubyte)*) s;
+    for (size_t i = 0; i != n; ++i)
```

Comment:
Why isn't this at least moving `ulong`'s?

Reply:
Your comment made me think, thanks.

I could use the modified GNU algorithm: https://www.embedded.com/print/4024961
but to me this only complicates the code without any real gain. The reason is that it only _tests_ if they're aligned and doesn't _reach_ alignment. To do a more serious 4-byte or 8-byte move, one has reach alignment etc. like this: https://github.com/dlang/druntime/pull/2662/files#diff-b2f13b9afdcf6d39739b2bf55035f8e2R166

The reason I did not do either of those is that the compiler can do this and other things itself. It should be able to understand that moving either 4 or 8 bytes at a time is faster, reach alignment, even add SIMD. For example
LDC: https://godbolt.org/z/2V1Wl8
GDC: https://godbolt.org/z/XXzLUQ (GDC needs -O3 to do the crazy stuff).

Here, the compiler does all of the above. It even added something like a switch.

But, here's the part I hadn't tested. If on LDC we pass the `-disable-loop-vectorization`
LDC: https://godbolt.org/z/WVVJlz
or we specify an architecture that does not support SIMD, like let's say `pentium2`.
LDC: https://godbolt.org/z/Q1vqOS
then it doesn't do 4-byte moves (It should be 4 bytes there since it is 32-bit compilation).

The same is true for GDC: https://godbolt.org/z/PK8YDQ

So, I don't know if it pays to _reach_ alignment but definetely I will use the GNU algorithm. It adds value without complicating the code much.

Reply:
You shouldn't be depending on optimisations that DMD can't do.
And LDC/GDC should be using intrinsics exclusively, definitely not calling any of these functions. Taking that knowledge away from the optimiser may create some very lame copies in optimised code.

Reply:
DMD was the initial focus. Then, it changed LDC and GDC so pretty much, what DMD does became irrelevant. Otherwise, the code would be way different. This is not a hypothesis. I spent one month writing DMD versions of these functions that had to reach the libc ones. They were completely different from the current ones due to the optimizations that DMD could not do.

Meaning, if LDC and GDC use their intrinsics, this project becomes irrelevant. But to use their intrinsics, the above question should be answered, which is what LDC and GDC do when libc is not available.

For me, the logical move is to inform the optimizer that e.g. our `memset` is supposed to do what `memset` does. Which enables us to write our own versions of these functions (aka libc independency) and not lose the optimization info.

Reply:
> which is what LDC and GDC do when libc is not available.

There's a difference between a call to a function called `mempy` (which is hooked by an intrinsic), and a symbol called `memcpy` in a lib somewhere. You're writing the latter, but the former is what appears in the code. LLVM/GCC intrinsics don't always emit a call to a memcpy function, they can do anything they like. They *might* emit a call to a memcpy function, and if they do, it would be an advantage to understand in what context that can occur.

I don't know the answer here; I haven't looked into this, it might be that they just call the libc memcpy when they decide not to do anything special; or they might call into their own small runtime libraries? Perhaps their runtime libraries have a small suite of such functions optimised for different cases? I don't know, but anything's possible here.
All the compilers have a runtime library which is a lower-level library than libc. I'm not sure if those libraries contain memcpy/memset's for when the intrinsics bail out... or if they all just leave it to the libc implementation? If the intrinsic's fall-back functions are in the runtime library and not taken from libc, then we already have libc independence in those cases.

Anyway, I can see a world where it may be useful to have these functions for when the intrinsic does bail-out, but I think it's important to understand the implementation details of each compilers intrinsics to understand what they expect from these functions and the conditions when they're called.

You'll never be able to write optimised implementations for all platforms (or even x86, because it's a moving target); so I feel like the most interesting victory here is for DMD.

I'm sorry, I don't really know anything about this project you're working on, or the parameters for completion... but I think tackling memcpy/memset are probably lower priority when considering libc independence, and possibly the 2 most highly contentious items on the list.

Reply:
> Meaning, if LDC and GDC use their intrinsics, this project becomes irrelevant.

Just to be clear; we should assume they are, and do absolutely everything we must to make sure we don't interfere with that. The case for us to consider here is when the intrinsics decide not to do anything special and fall-back to a lib-call... what functions do they call when they bail out? Is it libc memcpy/set, or something else customised for their runtimes?

Reply:
Musl may be a good reference too.

https://github.com/bminor/musl/blob/master/src/string/memcpy.c
https://github.com/bminor/musl/blob/master/src/string/memset.c

Reply:
@TurkeyMan I agree with everything you wrote. You see, when this project started (or when it moved to LDC / GDC), I didn't know any of these proeblems. It was Johan's comment (about a week ago) that made me realize "Oh, not using the intrinsics is like a really bad thing". I'm not sure how to move forward..

Reply:
@ibuclaw Yes, they're good reference, thanks. It was one of the early codes I was referred to. Generally though, Agner's implementations are supposed to be the fastest around (although that's not exactly true compared to the libc. This could be a very long discussion). He's considered a master in this field.

Other than that, unfortunately, I don't think that the main problems are with the code but with the goals.